baseapp/abci.go

```diff
@@ -289,6 +293,93 @@ func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) abci.Respon
 	return abci.ResponsePrepareProposal{Txs: txsBytes}
 }
 
+// Option1:
+// * Assume that app-side mempool will return valid set of txs. If it doesn't return a valid set then, it's considered a bug, so panic.
+// * `PrepareProposal` should perform lightweight validations only.
+func (app *BaseApp) PrepareProposalOption1(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+	var (
+		txsBytes  [][]byte
+		byteCount int64
+	)
+
+	ctx := app.getContextForTx(runTxPrepareProposal, []byte{})
+
+	// `Select` doesn't necessarily have to return an iterator. It can be a slice.
+	iterator := app.mempool.Select(ctx, req.Txs, req.MaxTxBytes)
```

Comment:
We opted for an iterator to accommodate highly customized cases like you're describing.  I'm leaning towards just letting apps define PrepareProposal.

Reply:
💯 I agree with this idea!